src/compiler/compiler.rs

```diff
@@ -881,33 +882,54 @@ where
     };
     let filename = filename.to_string_lossy().to_lowercase();
 
-    let rustc_vv = if filename == "rustc" || filename == "clippy-driver" {
-        // Sanity check that it's really rustc.
-        let executable = executable.to_path_buf();
-        let mut child = creator.clone().new_command_sync(executable);
-        child.env_clear().envs(ref_env(env)).args(&["-vV"]);
+    // Detect the secnario where cargo is used with sccache as a RUSTC_WRAPPER and another tool as a
+    // RUSTC_WORKSPACE_WRAPPER. In that case "rustc" will be the first argument rather than the command.
+    // As a guardrail against false positives, also check for the CARGO environment variable which
+    // cargo sets for any process it runs.
+    // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
+    let using_rustc_workspace_wrapper = env.iter().any(|(k, _)| k == OsStr::new("CARGO"))
+        && filename != "rustc"
+        && args.iter().next().map(|arg1| arg1.as_os_str()) == Some(OsStr::new("rustc"));
+
```

Comment:
Does it always have to be `rustc` or could it be something different as well? How do the mentioned
cargo tools integrate here?
I am mostly thinking for the upcoming changeset that's going to fix <https://github.com/mozilla/sccache/issues/861>

Reply:
No, it could be something different as cargo users can configure the rustc executable via https://doc.rust-lang.org/cargo/reference/config.html#buildrustc (or environment variable).

That will currently only work if the last part of the configured rustc is "rustc" (or clippy-driver). My change breaks that scenario by assuming "rustc" is the executable whenever a wrapper is used.

To resolve, I plan to update this check to see if the first argument is a path with a `file_name()` of rustc, and also store that argument for use as the executable below (instead of assuming "rustc"). Then #861 can subsequently update this check to support arbitrary rustc's. Does that sound reasonable?

Reply:
I've done that now, which handles more rustc wrapping scenarios. #861 can update the `is_rustc_like` function to detect/use the proposed environment variable.